README
NAME
    `Alien::libtermkey' - Alien wrapping for libtermkey

DESCRIPTION
    This CPAN distribution wraps the C library libtermkey in a wrapper
    suitable to drive CPAN and other Perl-related build infrastructure.

    If the C library is already installed and known by pkg-config, this
    module provides a simple access to its configuration. If not, the
    process of installing it will install a locally-bundled copy of the
    library into perl's arch-specific library directory.

    This module bundles libtermkey version 0.16.

METHODS
    This module behaves like ExtUtils::PkgConfig, responding to the same
    methods, except that the module name is implied. Thus, the configuration
    can be obtained by calling

     $cflags = Alien::libtermkey->cflags
     $libs = Alien::libtermkey->libs

     $ok = Alien::libtermkey->atleast_version( $version )

     etc...

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Alien-libtermkey

0.16    2014/01/04 22:26:26
        [CHANGES]
         * Allow wrapping of existing system library or installing bundled one

0.15    CHANGES:
         * Try to detect GNU make and GNU libtool by probing, rather than
           hardcoding what is the likely command on each OS type.

0.14    CHANGES:
         * A better implementation of the three 'check' style veersion testing
           ExtUtils::PkgConfig methods

0.13    CHANGES:
         * Wrap libtermkey 0.16

0.12    BUGFIXES:
         * Fixes for cygwin:
            + GNU make is just called "make"
            + Close written .pc file before trying to rename() it

0.11    BUGFIXES:
         * Removed dead 'use ExtUtils::PkgConfig' line

0.10    BUGFIXES:
         * Don't hardcode libdir into .pm file, discover it by @INC search
         * Rewrite file paths in .pc file relative to libdir - should allow
           for relocatable installs that work both from blib/ and real
           installation location

0.09    BUGFIXES:
         * Bake real installation libdir into build .pm file rather than
           build-time blib/ dir

0.08    CHANGES:
         * Wrap libtermkey 0.15b

        BUGFIXES:
         * Declare configure_requires correctly

0.07    CHANGES:
         * Wrap libtermkey 0.15a

0.06    BUGFIXES:
         * Pack the /correct/ libtermkey distribution tarball

0.05    CHANGES:
         * File::Path::make_path needs File::Path 2.07
         * GNU make is called 'gmake' on most non-Linux platforms

0.04    CHANGES:
         * Wrap libtermkey 0.14
         * Run bundled libtermkey unit tests at ./Build test time

0.03    CHANGES:
         * Wrap libtermkey 0.13
         * Don't render wrapped module version into generated .pm file as it
           breaks online docs like search.cpan.org or metacpan.org

0.02    CHANGES
         * ExtUtils::PkgConfig is a confiure-time requirement
         * Build into blib/ then use M::B install
         * Render wrapped module version into generated .pm file

0.01    First version, released on an unsuspecting world.



lib/Alien/libtermkey.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2011-2014 -- leonerd@leonerd.org.uk

package Alien::libtermkey;

our $VERSION = '0.16';

use POSIX qw( WEXITSTATUS );

my $module = '@PKGCONFIG_MODULE@';
my $use_bundled = '@USE_BUNDLED@';

# libdir is the first @INC path that contains a pkgconfig/ dir
my $libdir;
if( $use_bundled ) {
   foreach my $inc ( @INC ) {
      $libdir = $inc and last if -d "$inc/pkgconfig";
   }
   defined $libdir or die "Cannot find my libdir containing pkgconfig";
}

=head1 NAME

C<Alien::libtermkey> - L<Alien> wrapping for F<libtermkey>

=head1 DESCRIPTION

This CPAN distribution wraps the C library F<libtermkey> in a wrapper suitable
to drive CPAN and other Perl-related build infrastructure.

If the C library is already installed and known by F<pkg-config>, this module
provides a simple access to its configuration. If not, the process of
installing it will install a locally-bundled copy of the library into perl's
arch-specific library directory.

This module bundles F<libtermkey> version 0.16.

=head1 METHODS

This module behaves like L<ExtUtils::PkgConfig>, responding to the same
methods, except that the module name is implied. Thus, the configuration can
be obtained by calling

 $cflags = Alien::libtermkey->cflags
 $libs = Alien::libtermkey->libs

 $ok = Alien::libtermkey->atleast_version( $version )

 etc...

=cut

my %check_methods = map { $_ => 1 } qw(
   atleast_version
   exact_version
   max_version
);

# I AM EVIL
sub AUTOLOAD
{
   our $AUTOLOAD =~ s/^.*:://;
   return defined _get_pkgconfig( $AUTOLOAD, @_ ) if $check_methods{$AUTOLOAD};
   return _get_pkgconfig( $AUTOLOAD, @_ );
}

sub _get_pkgconfig
{
   my ( $method, $self, @args ) = @_;

   $method =~ s/_/-/g;

   unshift @args, "--$method";

   local $ENV{PKG_CONFIG_PATH} = "$libdir/pkgconfig/" if $use_bundled;
   unshift @args, "--define-variable=libdir=$libdir" if $use_bundled;

   open my $eupc, "-|", "pkg-config", @args, $module or
      die "Cannot popen pkg-config - $!";

   my $ret = do { local $/; <$eupc> }; chomp $ret;
   close $eupc;

   return undef if WEXITSTATUS($?);
   return $ret;
}

sub libs
{
   # Append RPATH so that runtime linking actually works
   return _get_pkgconfig( libs => @_ ) . ( $use_bundled ? " -Wl,-R$libdir" : "" );
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


inc/Alien/make/Module/Build.pm
package Alien::make::Module::Build;

use strict;
use warnings;

use base qw( Module::Build );

use File::Basename qw( dirname );
use File::Spec;
use File::Path 2.07 qw( make_path );

use constant SRCDIR => "src";

# GNU make is called 'gmake' on most non-Linux platforms, gnumake on Dariwn
use constant MAKE => ( $^O eq "linux" )  ? "make" :
                     ( $^O eq "darwin" ) ? "gnumake" :
                                           "gmake";

# GNU libtool is called 'glibtool' on Darwin
use constant MAKEARGS => ( ( $^O eq "darwin" ) ? ( "LIBTOOL=glibtool" ) : () );

__PACKAGE__->add_property( 'tarball' );
__PACKAGE__->add_property( 'pkgconfig_module' );

sub _srcdir
{
   my $self = shift;
   return File::Spec->catdir( $self->base_dir, SRCDIR );
}

sub _stampfile
{
   my $self = shift;
   my ( $name ) = @_;

   return File::Spec->catfile( $self->base_dir, ".$name-stamp" );
}

sub in_srcdir
{
   my $self = shift;

   chdir( $self->_srcdir ) or
      die "Unable to chdir to srcdir - $!";

   shift->();
}

sub make_in_srcdir
{
   my $self = shift;
   my @args = @_;

   $self->in_srcdir( sub {
      system( MAKE, MAKEARGS, @args ) == 0 or
         die "Unable to make - $!";
   } );
}

sub ACTION_src
{
   my $self = shift;

   -d $self->_srcdir and return;

   my $tarball = $self->tarball;

   system( "tar", "xzf", $tarball ) == 0 or
      die "Unable to untar $tarball - $!";

   ( my $untardir = $tarball ) =~ s{\.tar\.[a-z]+$}{};

   -d $untardir or
      die "Expected to find a directory called $untardir\n";

   rename( $untardir, $self->_srcdir ) or
      die "Unable to rename src dir - $!";
}

sub ACTION_code
{
   my $self = shift;

   my $blib = File::Spec->catdir( $self->base_dir, "blib" );

   my $libdir = File::Spec->catdir( $blib, "arch" );
   my $incdir = File::Spec->catdir( $libdir, "include" );
   my $mandir = File::Spec->catdir( $blib, "libdoc" );

   my $pkgconfig_module = $self->pkgconfig_module;

   my $buildstamp = $self->_stampfile( "build" );

   unless( -f $buildstamp ) {
      $self->depends_on( 'src' );

      $self->make_in_srcdir( () );

      $self->make_in_srcdir( "install",
         "LIBDIR=$libdir",
         "INCDIR=$incdir",
         "MAN3DIR=$mandir",
         "MAN7DIR=$mandir",
      );

      open( my $stamp, ">", $buildstamp ) or die "Unable to touch .build-stamp file - $!";

      # Now hack the .pc file to put ${libdir}-relative paths in it
      my $pcfile = "$libdir/pkgconfig/$pkgconfig_module.pc";
      if( -f $pcfile ) {
         open my $in, "<", $pcfile or die "Cannot open $pcfile for reading - $!";
         open my $out, ">", "$pcfile.new" or die "Cannot open $pcfile.new for writing - $!";

         while( <$in> ) {
            s{\Q$libdir\E}{\${libdir}}g;
            print { $out } $_;
         }

         rename "$pcfile.new", $pcfile or die "Cannot rename $pcfile.new to $pcfile - $!";
      }
   }

   my @module_file = split m/::/, $self->module_name . ".pm";
   my $srcfile = File::Spec->catfile( $self->base_dir, "lib", @module_file );
   my $dstfile = File::Spec->catfile( $blib, "lib", @module_file );

   unless( $self->up_to_date( $srcfile, $dstfile ) ) {

      my %replace = (
         PKGCONFIG_MODULE => $pkgconfig_module,
      );

      # Turn ' into \' in replacements
      s/'/\\'/g for values %replace;

      $self->cp_file_with_replacement(
         srcfile => $srcfile,
         dstfile => $dstfile,
         replace => \%replace,
      );
   }
}

sub cp_file_with_replacement
{
   my $self = shift;
   my %args = @_;

   my $srcfile = $args{srcfile};
   my $dstfile = $args{dstfile};
   my $replace = $args{replace};

   make_path( dirname( $dstfile ), 0, 0777 );

   open( my $inh,  "<", $srcfile ) or die "Cannot read $srcfile - $!";
   open( my $outh, ">", $dstfile ) or die "Cannot write $dstfile - $!";

   while( my $line = <$inh> ) {
      $line =~ s/\@$_\@/$replace->{$_}/g for keys %$replace;
      print $outh $line;
   }
}

sub ACTION_test
{
   my $self = shift;

   $self->depends_on( "code" );

   $self->make_in_srcdir( "test" );
}

sub ACTION_clean
{
   my $self = shift;

   if( -d $self->_srcdir ) {
      $self->make_in_srcdir( "clean" );
   }

   unlink( $self->_stampfile( "build" ) );

   $self->SUPER::ACTION_clean;
}

sub ACTION_realclean
{
   my $self = shift;

   if( -d $self->_srcdir ) {
      system( "rm", "-rf", $self->_srcdir ); # best effort; ignore failure
   }

   $self->SUPER::ACTION_realclean;
}

0x55AA;
